File: solution/day20/src/main/java/domain/yahtzee/original/YahtzeeCalculator.java
package domain.yahtzee.original;

import io.vavr.Function1;
import io.vavr.collection.Map;
import io.vavr.collection.Seq;
import io.vavr.collection.Traversable;

import static io.vavr.collection.List.ofAll;

public class YahtzeeCalculator {
    private static final int ROLL_LENGTH = 5;
    private static final int MINIMUM_DIE = 1;
    private static final int MAXIMUM_DIE = 6;

    public static int number(int[] dice, int number) {
        return calculate(d -> d.filter(die -> die == number)
                        .sum()
                        .intValue()
                , dice);
    }

    public static int threeOfAKind(int[] dice) {
        return calculateNOfAKind(dice, 3);
    }

    public static int fourOfAKind(int[] dice) {
        return calculateNOfAKind(dice, 4);
    }

    public static int yahtzee(int[] dice) {
        return calculate(d -> hasNOfAKind(d, 5) ? Scores.YAHTZEE_SCORE : 0, dice);
    }

    private static int calculateNOfAKind(int[] dice, int n) {
        return calculate(d -> hasNOfAKind(d, n)
                        ? d.sum().intValue()
                        : 0,
                dice);
    }

    public static int fullHouse(int[] dice) {
        return calculate(d -> {
                    var dieFrequency = groupDieByFrequency(d);
                    return (dieFrequency.containsValue(3) && dieFrequency.containsValue(2)) ? Scores.HOUSE_SCORE : 0;
                }, dice
        );
    }

    public static int largeStraight(int[] dice) {
        return calculate(d -> d.sorted()
                .sliding(2)
                .forAll(pair -> pair.get(0) + 1 == pair.get(1)) ? Scores.LARGE_STRAIGHT_SCORE : 0, dice);
    }

    public static int smallStraight(int[] dice) {
        return calculate(d -> {
            var diceString = toSortedString(d);
            return isSmallStraight(diceString) ? 30 : 0;
        }, dice);
    }

    private static boolean isSmallStraight(String diceString) {
        return diceString.contains("1234") || diceString.contains("2345") || diceString.contains("3456");
    }

    private static boolean hasNOfAKind(Seq<Integer> dice, int n) {
        return groupDieByFrequency(dice)
                .values()
                .exists(count -> count >= n);
    }

    private static String toSortedString(Seq<Integer> dice) {
        return dice.sorted()
                .distinct()
                .mkString();
    }

    public static int chance(int[] dice) {
        return calculate(d -> d.sum().intValue(), dice);
    }

    private static Map<Integer, Integer> groupDieByFrequency(Seq<Integer> dice) {
        return dice
                .groupBy(x -> x)
                .mapValues(Traversable::length);
    }

    private static int calculate(Function1<Seq<Integer>, Integer> compute, int[] dice) {
        validateRoll(dice);
        return compute.apply(ofAll(dice));
    }

    private static void validateRoll(int[] dice) {
        if (hasInvalidLength(dice)) {
            throw new IllegalArgumentException("Invalid dice... A roll should contain 6 dice.");
        }

        if (containsInvalidDie(dice)) {
            throw new IllegalArgumentException("Invalid die value. Each die must be between 1 and 6.");
        }
    }

    private static boolean hasInvalidLength(int[] dice) {
        return dice == null || dice.length != ROLL_LENGTH;
    }

    private static boolean containsInvalidDie(int[] dice) {
        return ofAll(dice).exists(YahtzeeCalculator::isValidDie);
    }

    private static boolean isValidDie(int die) {
        return die < MINIMUM_DIE || die > MAXIMUM_DIE;
    }

    private static class Scores {
        private static final int YAHTZEE_SCORE = 50;
        private static final int HOUSE_SCORE = 25;
        private static final int LARGE_STRAIGHT_SCORE = 40;
    }

    private YahtzeeCalculator() {
    }
}

File: solution/ts/day20/src/domain/yahtzee/original/yahtzeeCalculator.ts
export class YahtzeeCalculator {
    private static readonly ROLL_LENGTH = 5;
    private static readonly MINIMUM_DIE = 1;
    private static readonly MAXIMUM_DIE = 6;

    static number = (dice: number[], number: number): number => this.calculate(
        d => d.filter(die => die === number)
            .reduce((sum, die) => sum + die, 0),
        dice
    );

    static threeOfAKind = (dice: number[]): number => this.calculateNOfAKind(dice, 3);

    static fourOfAKind = (dice: number[]): number => this.calculateNOfAKind(dice, 4);

    static yahtzee = (dice: number[]): number => this.calculate(
        d => d.every(die => die === dice[0])
            ? Scores.YAHTZEE_SCORE
            : 0,
        dice
    );

    static fullHouse = (dice: number[]): number => this.calculate(
        d => {
            const frequencies = this.getFrequencies(d);
            const hasThreeOfAKind = Object.values(frequencies).includes(3);
            const hasPair = Object.values(frequencies).includes(2);

            return hasThreeOfAKind && hasPair
                ? Scores.HOUSE_SCORE
                : 0;
        },
        dice
    );

    static smallStraight = (dice: number[]): number => this.calculate(
        d => {
            const sortedUniqueDice = Array.from(new Set(d)).sort();
            const diceString = sortedUniqueDice.join('');

            return this.isSmallStraight(diceString)
                ? 30
                : 0;
        },
        dice
    );

    static largeStraight = (dice: number[]): number => this.calculate(
        d => {
            const sortedDice = [...d].sort((a, b) => a - b);
            const isSequential = sortedDice.every((die, index, arr) => index === 0 || die === arr[index - 1] + 1);

            return isSequential
                ? Scores.LARGE_STRAIGHT_SCORE
                : 0;
        },
        dice
    );

    static chance = (dice: number[]): number => this.calculate(
        d => d.reduce((sum, die) => sum + die, 0),
        dice
    );

    private static calculateNOfAKind = (dice: number[], n: number): number => this.calculate(
        d => {
            const hasNOfAKind = Object.values(this.getFrequencies(d)).some(count => count >= n);
            return hasNOfAKind
                ? d.reduce((sum, die) => sum + die, 0)
                : 0;
        },
        dice
    );

    private static validateRoll(dice: number[]): void {
        if (dice.length !== this.ROLL_LENGTH) {
            throw new Error('Invalid dice... A roll should contain 5 dice.');
        }

        if (dice.some(die => die < this.MINIMUM_DIE || die > this.MAXIMUM_DIE)) {
            throw new Error('Invalid die value. Each die must be between 1 and 6.');
        }
    }

    private static getFrequencies = (dice: number[]): Record<number, number> =>
        dice.reduce((freq, die) => {
            freq[die] = (freq[die] || 0) + 1;
            return freq;
        }, {} as Record<number, number>);

    private static isSmallStraight = (diceString: string): boolean =>
        ['1234', '2345', '3456'].some(straight => diceString.includes(straight));

    private static calculate(compute: (dice: number[]) => number, dice: number[]): number {
        this.validateRoll(dice);
        return compute(dice);
    }
}

const Scores = {
    YAHTZEE_SCORE: 50,
    HOUSE_SCORE: 25,
    LARGE_STRAIGHT_SCORE: 40
};